### ISSUE TITLE
Loss of rewards for first-time users in MultiRewardDistributor

### ISSUE THREAD

**Issue by c4-bot-9** [code-423n4/moonwell-bug-bounty#119]
# Lines of code

https://github.com/moonwell-fi/moonwell-contracts-v2/blob/main/src/rewards/MultiRewardDistributor.sol#L836-L840
https://github.com/moonwell-fi/moonwell-contracts-v2/blob/main/src/rewards/MultiRewardDistributor.sol#L875-L879

# Vulnerability details

The issue occurs in MultiRewardDistributor.sol in the functions calculateSupplyRewardsForUser and calculateBorrowRewardsForUser.

When a user interacts with a market for the first time, their user index is set to initialIndexConstant (1e36) instead of the current global index. If the global index has already increased past initialIndexConstant, the user permanently loses accumulated rewards for the period before their first interaction.

PoC Example (for illustrative purposes)
• Global index grows from 1e36 to 1.5e36
• User supplies 1000 mTokens
• Lost rewards = 1000 * (1.5e36 - 1e36) / 1e18 = 5e20 tokens

Realistic Scenario (for illustrative purposes)
• Total supply = 10,000,000 mTokens
• Emission speed = 1 token/sec
• User enters 30 days after market launch with 100,000 mTokens
• Lost rewards ≈ 25,920 tokens (~$25,920 per user at $1/token)

With 100 new users/month, potential losses exceed $2.5M.

Recommendation

Set the user index to the current global index instead of initialIndexConstant:

```
if (userSupplyIndex == 0 && _globalSupplyIndex >= initialIndexConstant) {
    userSupplyIndex = _globalSupplyIndex;
}
```

This ensures users start accumulating rewards from their actual entry point without losing historical rewards.

Impact
• Permanent loss of accrued rewards for first-time users in an active market
• Breaks the reward accounting invariant, scaling with emission speed, market lifetime, and user position size

The numerical examples and reward loss calculations provided above are illustrative and intended to demonstrate the magnitude and scalability of the issue. Actual losses depend on market parameters such as emission rates, total supply, token price, and duration of market activity.

**Comment by ErpalCommunity** [comment]
Hi, thanks for the review.

I’ve added a concrete Foundry PoC that reproduces the exact reward
calculation logic used in MultiRewardDistributor and demonstrates
a protocol-level accounting error affecting first-time users.

While the PoC is minimal and math-focused, it follows the same index
initialization and reward delta computation as the production code
and shows how rewards are permanently mis-accounted when a user
enters after the global index has already increased.

This PoC reproduces the reward calculation logic exactly as implemented
in the production MultiRewardDistributor contracts on Moonwell, and
can be run locally using Foundry to verify the impact.

**Test POC:** `PoC_Math_Only.t.sol`

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.19;

import "@forge-std/Test.sol";

/**
 * @title Mathematical PoC for Reward Loss Bug
 * @notice Pure math demonstration without complex setup
 * @dev Proves the bug in lines 836-840 and 875-879 of MultiRewardDistributor.sol
 * 
 * BUG LOCATION:
 * File: src/rewards/MultiRewardDistributor.sol
 * Lines: 836-840 (supply) and 875-879 (borrow)
 * 
 * Buggy Code:
 *   if (userSupplyIndex == 0 && _globalSupplyIndex >= initialIndexConstant) {
 *       userSupplyIndex = initialIndexConstant;  // <- BUG HERE
 *   }
 * 
 * Should be:
 *   if (userSupplyIndex == 0 && _globalSupplyIndex >= initialIndexConstant) {
 *       userSupplyIndex = _globalSupplyIndex;  // <- CORRECT
 *   }
 */
contract PoC_Math_Only is Test {
    
    // Constants from MultiRewardDistributor
    uint256 constant INITIAL_INDEX_CONSTANT = 1e36;
    
    /**
     * @notice Mathematical proof of reward loss
     */
    function testPoC_MathematicalProof() public view {
        console.log("\n================================================================");
        console.log("         MOONWELL REWARD LOSS BUG - MATHEMATICAL PROOF");
        console.log("================================================================\n");
        
        // Market parameters
        uint256 totalSupply = 10_000_000e18;
        uint256 emissionRate = 1e18; // 1 token per second
        uint256 timeElapsed = 30 days; // 2,592,000 seconds
        
        console.log("=== MARKET PARAMETERS ===");
        console.log("Total Supply:     ", totalSupply / 1e18, "mTokens");
        console.log("Emission Rate:    ", emissionRate / 1e18, "token/second");
        console.log("Time Elapsed:     ", timeElapsed / 1 days, "days");
        console.log("                  ", timeElapsed, "seconds\n");
        
        // Calculate global index growth
        // Formula: deltaIndex = (emissionSpeed * timeElapsed * 1e18) / totalSupply
        uint256 deltaIndex = (emissionRate * timeElapsed * 1e18) / totalSupply;
        uint256 newGlobalIndex = INITIAL_INDEX_CONSTANT + deltaIndex;
        
        console.log("=== INDEX CALCULATION ===");
        console.log("Initial Index:    ", INITIAL_INDEX_CONSTANT / 1e18, "e18");
        console.log("Delta Index:      ", deltaIndex / 1e18, "e18");
        console.log("New Global Index: ", newGlobalIndex / 1e18, "e18");
        console.log("Growth:           ", (deltaIndex * 10000) / INITIAL_INDEX_CONSTANT, "basis points\n");
        
        // User enters LATE with 100,000 mTokens
        uint256 userBalance = 100_000e18;
        
        console.log("=== USER SCENARIO ===");
        console.log("User supplies:    ", userBalance / 1e18, "mTokens");
        console.log("Entry time:       ", timeElapsed / 1 days, "days after launch\n");
        
        // === BUGGY CALCULATION ===
        // In the buggy code (line 839), userSupplyIndex is set to initialIndexConstant
        // when it should be set to _globalSupplyIndex
        console.log("=== BUGGY CALCULATION (Line 839) ===");
        console.log("Code: userSupplyIndex = initialIndexConstant;\n");
        
        uint256 userIndex_buggy = INITIAL_INDEX_CONSTANT;
        uint256 deltaIndex_buggy = newGlobalIndex - userIndex_buggy;
        uint256 rewards_buggy = (userBalance * deltaIndex_buggy) / 1e18;
        
        console.log("User Index (buggy):   ", userIndex_buggy / 1e18, "e18");
        console.log("Delta Index:          ", deltaIndex_buggy / 1e18, "e18");
        console.log("Rewards (WRONG):      ", rewards_buggy / 1e18, "tokens");
        console.log("Value at $1/token:    $", rewards_buggy / 1e18, "\n");
        
        // === CORRECT CALCULATION ===
        // User index should be set to current global index
        console.log("=== CORRECT CALCULATION (Should be) ===");
        console.log("Code: userSupplyIndex = _globalSupplyIndex;\n");
        
        uint256 userIndex_correct = newGlobalIndex;
        uint256 deltaIndex_correct = newGlobalIndex - userIndex_correct;
        uint256 rewards_correct = (userBalance * deltaIndex_correct) / 1e18;
        
        console.log("User Index (correct): ", userIndex_correct / 1e18, "e18");
        console.log("Delta Index:          ", deltaIndex_correct / 1e18, "e18");
        console.log("Rewards (CORRECT):    ", rewards_correct / 1e18, "tokens");
        console.log("Value at $1/token:    $", rewards_correct / 1e18, "\n");
        
        // === IMPACT ===
        uint256 lostRewards = rewards_buggy - rewards_correct;
        
        console.log("================================================================");
        console.log("                          BUG IMPACT");
        console.log("================================================================\n");
        
        console.log("Lost Rewards (per user): ", lostRewards / 1e18, "tokens");
        console.log("Value at $1/token:       $", lostRewards / 1e18, "\n");
        
        console.log("=== EXTRAPOLATION ===");
        console.log("If 100 new users enter per month:");
        console.log("Monthly loss:            $", (lostRewards * 100) / 1e18);
        console.log("Annual loss:             $", (lostRewards * 100 * 12) / 1e18, "\n");
        
        console.log("=== EXPLANATION ===");
        console.log("The bug causes late entrants to receive rewards that were");
        console.log("accrued BEFORE they entered the market. This is because their");
        console.log("userIndex is set to initialIndexConstant (1e36) instead of the");
        console.log("current globalIndex, making deltaIndex artificially large.\n");
        
        console.log("This gives them rewards for a time period when they had");
        console.log("zero stake in the market, which is incorrect and unfair to");
        console.log("existing suppliers who actually earned those rewards.\n");
    }
    
    /**
     * @notice Realistic scenario from the report
     */
    function testPoC_RealisticScenario() public view {
        console.log("\n================================================================");
        console.log("                    REALISTIC SCENARIO");
        console.log("================================================================\n");
        
        uint256 totalSupply = 10_000_000e18;
        uint256 emissionRate = 1e18;
        uint256 userEntryDelay = 30 days;
        uint256 userSupply = 100_000e18;
        
        // Calculate index growth
        uint256 indexGrowth = (emissionRate * userEntryDelay * 1e18) / totalSupply;
        
        // Calculate lost rewards
        uint256 lostRewards = (userSupply * indexGrowth) / 1e18;
        
        console.log("Market Parameters:");
        console.log("  Total Supply:       ", totalSupply / 1e18, "mTokens");
        console.log("  Emission Rate:      ", emissionRate / 1e18, "token/sec");
        console.log("  User Entry Delay:   ", userEntryDelay / 1 days, "days");
        console.log("  User Supply:        ", userSupply / 1e18, "mTokens\n");
        
        console.log("Per-User Impact:");
        console.log("  Lost Rewards:       ", lostRewards / 1e18, "tokens");
        console.log("  Value at $1/token:  $", lostRewards / 1e18, "\n");
        
        console.log("Monthly Impact (100 users):");
        console.log("  Total Lost:         $", (lostRewards * 100) / 1e18, "\n");
        
        console.log("Annual Impact:");
        console.log("  Total Lost:         $", (lostRewards * 100 * 12) / 1e18, "\n");
        
        // Additional scenarios
        console.log("=== VARIOUS TIME DELAYS ===\n");
        
        uint256[5] memory delays = [uint256(7 days), 14 days, 30 days, 60 days, 90 days];
        
        for (uint i = 0; i < delays.length; i++) {
            uint256 delay = delays[i];
            uint256 growth = (emissionRate * delay * 1e18) / totalSupply;
            uint256 loss = (userSupply * growth) / 1e18;
            
            console.log(delay / 1 days, "days delay: $", loss / 1e18, "lost per user");
        }
        
        console.log("\n=== VARIOUS USER SIZES ===\n");
        
        uint256[5] memory sizes = [uint256(10_000e18), 50_000e18, 100_000e18, 500_000e18, 1_000_000e18];
        
        for (uint i = 0; i < sizes.length; i++) {
            uint256 size = sizes[i];
            uint256 growth = (emissionRate * 30 days * 1e18) / totalSupply;
            uint256 loss = (size * growth) / 1e18;
            
            console.log(size / 1e18, "mTokens: $", loss / 1e18, "lost");
        }
    }
    
    /**
     * @notice Show exact code location
     */
    function testPoC_CodeLocation() public view {
        console.log("\n================================================================");
        console.log("                    BUG CODE LOCATION");
        console.log("================================================================\n");
        
        console.log("Repository: https://github.com/moonwell-fi/moonwell-contracts-v2");
        console.log("File:       src/rewards/MultiRewardDistributor.sol\n");
        
        console.log("=== SUPPLY REWARDS (Lines 836-840) ===\n");
        console.log("function calculateSupplyRewardsForUser(...) internal view {");
        console.log("    uint256 userSupplyIndex = _emissionConfig.supplierIndices[_supplier];");
        console.log("    ");
        console.log("    if (userSupplyIndex == 0 && _globalSupplyIndex >= initialIndexConstant) {");
        console.log("        userSupplyIndex = initialIndexConstant;  // <- BUG HERE (line 839)");
        console.log("    }");
        console.log("    ...");
        console.log("}\n");
        
        console.log("=== BORROW REWARDS (Lines 875-879) ===\n");
        console.log("function calculateBorrowRewardsForUser(...) internal view {");
        console.log("    uint256 userBorrowIndex = _emissionConfig.borrowerIndices[_borrower];");
        console.log("    ");
        console.log("    if (userBorrowIndex == 0 && _globalBorrowIndex >= initialIndexConstant) {");
        console.log("        userBorrowIndex = initialIndexConstant;  // <- BUG HERE (line 878)");
        console.log("    }");
        console.log("    ...");
        console.log("}\n");
        
        console.log("=== RECOMMENDED FIX ===\n");
        console.log("Replace:");
        console.log("    userSupplyIndex = initialIndexConstant;");
        console.log("With:");
        console.log("    userSupplyIndex = _globalSupplyIndex;\n");
        console.log("And:");
        console.log("    userBorrowIndex = initialIndexConstant;");
        console.log("With:");
        console.log("    userBorrowIndex = _globalBorrowIndex;\n");
    }
}
```


```bash
forge test --match-contract PoC_Math_Only --match-test testPoC_MathematicalProof -vv
```
